### PR TITLE
poc: Node drivers PoC 

### PR DESCRIPTION
**Pull request by miguelff**
Supersedes https://github.com/prisma/prisma-engines/pull/3973 in an integration branch with the goal of generating integrating binaries through our engineer pipeline.

@jkomyno's description of  https://github.com/prisma/prisma-engines/pull/3973  follows:

## Preliminaries

Tested through https://github.com/prisma/prisma-engines/pull/4006

## Description

This PR:
- Adds a new crate, `nodejs-drivers`, responsible for defining a `Queryable` + `TransactionCapable` + `Send` + `Sync` implementation that uses functions exposed by Node.js drivers via N-API. Closes https://github.com/prisma/team-orm/issues/10.
- Adds a new `fn_ctx: Option<JsObject>` argument to `QueryEngine`'s constructor in the `query-engine-node-api` crate. This is used to create a new instance of `NodeJSQueryable`, the new main class of `nodejs-drivers`. Closes https://github.com/prisma/team-orm/issues/11.
- Integrates `NodeJSQueryable` in `mysql` within the `sql-query-connector` crate. The database drivers used at runtime by Prisma now varies depending on the provider specified in the Prisma schema. Now, with the `mysql` provider, Prisma uses Quaint's MySQL drivers under the hood, whereas with `@prisma/mysql` provider it uses `NodeJSQueryable`. 
Closes https://github.com/prisma/team-orm/issues/9.
- Adds the `nodejs-drivers` Cargo feature, to dynamically compile `napi.rs` modules that may be transitively imported by `query-engine` and `connector-test-kit-rs/*` and ensure a green CI. Closes https://github.com/prisma/team-orm/issues/60.

## NodeJSQueryable's allocation

Initially, Miguel, Serhii, and I, were oriented for storing the `NodeJSQueryable` instance directly in the `QueryEngine` struct, wrapped in an `Option` that would become `None` after the first usage of `NodeJSQueryable` in the `QueryEngine::connect()` method.

However, I figured this could lead to a panic whenever `QueryEngine::connect()` is called for the second time after a `QueryEngine::disconnect()` (as can be seen in https://github.com/prisma/prisma-engines/pull/3973/commits/ef47ef4e08f18fb151eb5bc4de850ca4eef98880). I have applied two (alternative) fixes:
- one that clones the `Option<NodeJSQueryable>` before reading it (making the `RwLock` overkill, see https://github.com/prisma/prisma-engines/pull/3973/commits/325c80007ba3c573231056ee578c530e7499a15e)
- one that spawns a global `NodeJSQueryable` via `once_cell`, which Tom likely prefers anyway (https://github.com/prisma/prisma-engines/pull/3973/commits/d29fc5c82b05ae2dd588d599f8eac305ef8ab5d2).

I hope this clarifies the initial doubts that people from the Client team and I had before the discoveries made throughout the implementation.

[Board Issue](https://github.com/prisma/team-orm/issues/180)

### CODE REVIEW

**Review by jkomyno — APPROVED** [review]

File: query-engine/nodejs-drivers/src/driver.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        println!("[rs] calling query_raw: {}", &sql);
+
+        let promise = self.query_raw.call_async::<JsPromise<ResultSet>>(sql).await?;
+
+        println!("[rs] awaiting promise");
+        let value = promise.await?;
+
+        println!("[rs] awaited: {:?}", &value);
+        Ok(value)
+    }
```
Comment (jkomyno) [inline query-engine/nodejs-drivers/src/driver.rs:70]: Note: the `println!` invocations in this file will never alter the I/O of the "standard" query engine, and will be preserved here until we have are ready to publicly launch the `nodeDrivers` preview feature.

File: query-engine/query-engine-node-api/src/engine.rs [hunk 2]
```diff
@@ -138,11 +139,25 @@ impl Inner {
...
+    /// Note: any new method added to this struct should be added to
+    /// `query_engine_node_api::node_drivers::engine::QueryEngineNodeDrivers` as well.
+    /// Unfortunately the `#[napi]` macro does not support deriving traits.
     #[napi(constructor)]
-    pub fn new(napi_env: Env, options: JsUnknown, callback: JsFunction) -> napi::Result<Self> {
+    pub fn new(
+        napi_env: Env,
+        options: JsUnknown,
+        callback: JsFunction,
+        fn_ctx: Option<JsObject>,
```
Comment (jkomyno) [inline query-engine/query-engine-node-api/src/engine.rs:150]: If no `fn_ctx` is passed (i.e., if the 3rd parameter of the JS QueryEngine constructor is `undefined`), this is simply considered as a `None` value.

Comment (jkomyno) [reply query-engine/query-engine-node-api/src/engine.rs:150]: In other words, the previous TS logic for the QueryEngine can remain unaltered (until we're able to ship the `nodeDrivers` feature in preview).

**Comment by jkomyno** [comment]
Merging as agreed with Miguel on Friday.

**Comment by aqrln** [comment]
Building both `query-engine` and `query-engine-node-api` (e.g. by running `cargo build`) is currently broken on Linux on `main` after merging. This also affects certain CI jobs that only run on main ("Publish engine size" job now fails with linker errors). We'll have to fix this soon.

**Pull request merged**